### PR TITLE
feat: support postgres query hook

### PR DESCRIPTION
**Pull request by yihong0618**
Fix:#48   basic support for `Postgres 10+`  will do the `return value` in another `PR` if could

image

### CODE REVIEW

File: cli/cmd/postgres.go [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+//postgres Cmd represents the postgres command
+var postgresCmd = &cobra.Command{
+	Use:   "postgres",
+	Short: "capture sql queries from postgres 10+.",
+	Run:   postgresCommandFunc,
+}
+
+func init() {
+	postgresCmd.PersistentFlags().StringVarP(&postgresConfig.PostgresPath, "postgres", "m", "/usr/bin/postgres", "postgres binary file path, use to hook")
+	postgresCmd.PersistentFlags().Uint64VarP(&postgresConfig.Offset, "offset", "", 0, "0x710410")
```
Comment (cfc4n) [inline cli/cmd/postgres.go:29]: These values should not be constant. Why not use symbol name?

File: kern/postgres_kern.c [hunk 2]
```diff
@@ -0,0 +1,42 @@
+#include "ecapture.h"
+#include "common.h"
```
Comment (cfc4n) [inline kern/postgres_kern.c:2]: It was included in ecapture.h

Comment (yihong0618) [reply kern/postgres_kern.c:2]: copy that

File: cli/cmd/postgres.go (hunk 1 above)
Comment (yihong0618) [reply cli/cmd/postgres.go:29]: will take a look and change it ~

**Review by cfc4n — COMMENTED** [review]
How to know that It works with postgres binrary? Can you give some screenshots of catrure with postgres?

File: kern/postgres_kern.c [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+{
+    __uint(type, BPF_MAP_TYPE_PERF_EVENT_ARRAY);
+} events SEC(".maps");
+
+// https://github.com/postgres/postgres/blob/7b7ed046cb2ad9f6efac90380757d5977f0f563f/src/backend/tcop/postgres.c#L987-L992
+// hook function exec_simple_query
+// versions 10 - now
+// static void exec_simple_query(const char *query_string)
+SEC("uprobe/exec_simple_query")
+int postgres_query(struct pt_regs *ctx) {
```
Comment (cfc4n) [inline kern/postgres_kern.c:21]: what is the difference between `exec_simple_query` and `exec_parse_message`, Why not `exec_parse_message`?

Comment (yihong0618) [reply kern/postgres_kern.c:21]: exec_parse_message called in exec_simple_query.

**Review by cfc4n — COMMENTED** [review]
portgres use two function to exec queries in  https://github.com/postgres/postgres/blob/7b7ed046cb2ad9f6efac90380757d5977f0f563f/src/backend/tcop/postgres.c#L4539  And https://github.com/postgres/postgres/blob/7b7ed046cb2ad9f6efac90380757d5977f0f563f/src/backend/tcop/postgres.c#L4571  whit params `Q` and `P`,  why is that ? and this PR can capture all query commands?

**Comment by yihong0618** [comment]
> How to know that It works with postgres binrary? Can you give some screenshots of catrure with postgres?

Done, the screenshoot in the content now.

**Comment by yihong0618** [comment]
> portgres use two function to exec queries in https://github.com/postgres/postgres/blob/7b7ed046cb2ad9f6efac90380757d5977f0f563f/src/backend/tcop/postgres.c#L4539 And https://github.com/postgres/postgres/blob/7b7ed046cb2ad9f6efac90380757d5977f0f563f/src/backend/tcop/postgres.c#L4571 whit params `Q` and `P`, why is that ? and this PR can capture all query commands?

yes.

**Comment by yihong0618** [comment]
> portgres use two function to exec queries in https://github.com/postgres/postgres/blob/7b7ed046cb2ad9f6efac90380757d5977f0f563f/src/backend/tcop/postgres.c#L4539 And https://github.com/postgres/postgres/blob/7b7ed046cb2ad9f6efac90380757d5977f0f563f/src/backend/tcop/postgres.c#L4571 whit params `Q` and `P`, why is that ? and this PR can capture all query commands?

yes, it capture all the query I think its the same logic as mysqld

File: cli/cmd/postgres.go (hunk 1 above)
Comment (yihong0618) [reply cli/cmd/postgres.go:29]: And I think I should drop `offset`

**Comment by yihong0618** [comment]
@cfc4n Done and clean some code.

**Comment by cfc4n** [comment]
please merge all commits into one commit with command
```
git reset --soft HEAD~3
git commit
git push -f
```

**Review by cfc4n — COMMENTED** [review]
refer : https://www.internalpointers.com/post/squash-commits-into-one-git

**Comment by yihong0618** [comment]
Of course I did that in last `PR`. https://github.com/ehids/ecapture/pull/15#discussion_r837339156

And I think the the `repo` owner can use `squash and merge` which can also merge all commit into one.

**Comment by cfc4n** [comment]
LGTM, thanks.

**Comment by yihong0618** [comment]
> LGTM, thanks.

Will do some test on other env and add it to README

**Pull request merged**